k4FWCore/components/EfficiencyFilter.cpp

```diff
@@ -0,0 +1,138 @@
+
+    return StatusCode::SUCCESS;
+  }
+
+  podio::CollectionBase* operator()(const podio::CollectionBase& coll,
+                                    const edm4hep::EventHeaderCollection& evtHeader) const override {
+
+    debug() << "EfficiencyFilter called for collection of type " << coll.getTypeName() << endmsg;
+
+    const auto uid = m_uniqueIDSvc->getUniqueID(evtHeader, name());
+
+    auto ptr =
+        dispatchByType(coll, uid, edm4hep::edm4hepDataTypes{}); // Dispatch to the correct merge function based on type
+    if (!ptr) {
+      ptr = dispatchByType(coll, uid, edm4hep::edm4hepLinkTypes{}); // Try link types if not found in data types
+    }
+    if (!ptr) {
+      throw std::runtime_error(
+          fmt::format("EfficiencyFilter: No merge function found for collection type '{}'", coll.getTypeName()));
+    }
+    return ptr;
+  }
+
+private:
+  SmartIF<IUniqueIDGenSvc> m_uniqueIDSvc{nullptr};
+
+  Gaudi::Property<double> m_efficiency{this, "Efficiency", 0.5,
+                                       "The efficiency of the filter, between 0 and 1. If 0.5, each item has a 50% "
+                                       "chance to be kept in the output collection."};
+  Gaudi::Property<bool> m_exact{
+      this, "Exact", true,
+      "If true, the output collection will have exactly the number of elements given by the efficiency. For example, "
+      "if the efficiency is 0.5 and the input collection has 10 elements, the output collection will have exactly 5 "
+      "elements. "
+      "If false, the output collection will have a random number of elements, with the average number of elements "
+      "given by the efficiency."};
+
+  template <typename... T>
+  podio::CollectionBase* dispatchByType(const podio::CollectionBase& coll, size_t uid,
+                                        podio::utils::TypeList<T...>&&) const {
+    std::string_view typeName = coll.getTypeName();
+    podio::CollectionBase* result = nullptr;
+    (..., (typeName == T::collection_type::typeName
+               ? (result = createSubsetCollection<typename T::collection_type>(&coll, uid))
+               : nullptr));
+    return result;
+  }
+
+  template <typename T>
+  podio::CollectionBase* createSubsetCollection(const podio::CollectionBase* source, const size_t uid) const {
+
+    std::mt19937 randomGen(std::random_device{}());
+    randomGen.seed(uid);
+    std::uniform_real_distribution<double> uniformDist(0.0, 1.0);
+
+    auto ret = new T();
+    ret->setSubsetCollection();
+    const auto ptr = static_cast<T*>(ret);
+    const auto sourceColl = static_cast<const T*>(source);
+
+    if (m_exact) {
+      // If exact, we will keep exactly the number of elements given by the efficiency
+      const size_t nToKeep = static_cast<size_t>(static_cast<double>(sourceColl->size()) * m_efficiency);
+
+      std::vector<size_t> indexes(sourceColl->size());
+      std::iota(indexes.begin(), indexes.end(), 0);
+      std::ranges::shuffle(indexes, randomGen);
```

Comment:
Take the first `nToKeep` indices and sort them, so that the order of hits is the same as in the original.

Reply:
This is now done